File: estore/src/main/java/dev/a2/estore/model/Address.java
/*
 * MIT License
 *
 * Copyright (c) 2019 Andrei Sidorov
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dev.a2.estore.model;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;

/**
 * This model represents an address of a user.
 *
 * @author Andrei Sidorov
 */
@Entity
@Table(name = "addresses")
public class Address {

    /**
     * The id an address. Generated automatically.
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "address_id", unique = true, nullable = false)
    private Long id;

    /**
     * The country of an address.
     */
    @ManyToOne
    @JoinColumn(name = "country_id")
    private Country country;

    /**
     * The name of a city.
     */
    @Column(name = "city")
    private String city;

    /**
     * The postal code of an address.
     */
    @Column(name = "postalCode")
    private String postalCode;

    /**
     * The name of a street.
     */
    @Column(name = "street")
    private String street;

    /**
     * The number of a house.
     */
    @Column(name = "house")
    private String house;

    /**
     * The number of an apartment.
     */
    @Column(name = "apartment")
    private String apartment;

    /* Getters and setters */

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Country getCountry() {
        return country;
    }

    public void setCountry(Country country) {
        this.country = country;
    }

    public String getCity() {
        return city;
    }

    public void setCity(String city) {
        this.city = city;
    }

    public String getPostalCode() {
        return postalCode;
    }

    public void setPostalCode(String postalCode) {
        this.postalCode = postalCode;
    }

    public String getStreet() {
        return street;
    }

    public void setStreet(String street) {
        this.street = street;
    }

    public String getHouse() {
        return house;
    }

    public void setHouse(String house) {
        this.house = house;
    }

    public String getApartment() {
        return apartment;
    }

    public void setApartment(String apartment) {
        this.apartment = apartment;
    }

}

File: estore/src/main/resources/static/js/address.js
/*
 * MIT License
 *
 * Copyright (c) 2019 Andrei Sidorov
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
"use strict";

// Provides validation for user input, when input is valid sets input border color to green,
// when input is invalid sets border color to red and provides a custom validation message.
function validate(input) {
    let isValid, message;
    const inputValue = input.value.trim();
    const inputValueLength = inputValue.length;

    switch(input.id) {
        case "city":
            isValid = /^.{1,35}$/.test(inputValue);
            message = inputValueLength > 1 ? "Invalid city name" : "Please enter your city name";
            break;
        case "postalCode":
            isValid = /^.{1,16}$/.test(inputValue);
            message = inputValueLength > 1 ? "Invalid postal code" : "Please enter your postal code";
            break;
        case "street":
            isValid = /^.{1,35}$/.test(inputValue);
            message = inputValueLength > 1 ? "Invalid street name" : "Please enter your street name";
            break;
        case "house":
            isValid = /^.{1,16}$/.test(inputValue);
            message = inputValueLength > 1 ? "Invalid house number" : "Please enter your house number";
            break;
        case "apartment":
            isValid = /^.{1,16}$/.test(inputValue);
            message = inputValueLength > 1 ? "Invalid apartment number" : "Please enter your apartment number";
            break;
    }

    if (isValid) {
        input.style.borderColor = "green";
        input.setCustomValidity("");
        return true;
    } else {
        input.style.borderColor = "red";
        input.setCustomValidity(message);
        return false;
    }

}

// Sets the red border and provides error messsage to the country field if a country is not chosen
const countryInput = document.getElementById("countryId");
if (countryInput.value === "Choose your country") {
    countryInput.setCustomValidity("Please choose your country");
}

countryInput.addEventListener("focusout", () => {
    if (countryInput.value !== "Choose your country") {
        countryInput.style.borderColor = "green";
        countryInput.setCustomValidity("");
    } else {
        countryInput.style.borderColor = "red";
        countryInput.setCustomValidity("Please choose your country");
    }
});

// Provides red border for the email field if a user with the entered email exists.
const errors = document.querySelectorAll(".error");
for (let i = 0; i < errors.length; i++) {
    const input = errors[i].parentElement.querySelector("#email");
    input.style.borderColor = "red";

}


